### PR TITLE
[data] Add databricks table / SQL reader

### PR DESCRIPTION
**Pull request by WeichenXu123**
## Why are these changes needed?

Databricks users request a `ray.data` reader that can easily load data from databricks UC tables.
This reader is implemented based on [Databricks statemenet execution API](https://docs.databricks.com/api/workspace/statementexecution), the overall process is:

1. `read_databricks_tables` create request `/api/2.0/sql/statements` to databricks shard, request payload contains authentication token and query information.
2. get response that contains the query result chunks information. The query result is split into multiple chunks, each chunk data size is around 10 ~ 20 MB.
3. Generate a couple of Ray Read tasks, read tasks number is determined by `parallelism` argument. Each read task will fetch one or more chunks data. Feching chunks data uses this [API](https://docs.databricks.com/api/workspace/statementexecution/getstatementresultchunkn).
4. Ray read tasks are dispatched to remote Ray workers and executed in distributed way.

API:

```python
    """
    Read from a Databricks UC table or Databricks SQL execution result that queries
    from Databricks UC tables.
    If it is not called in databricks runtime, you need to set environment varaibles
    'DATABRICKS_HOST' and 'DATABRICKS_TOKEN' firstly.

    This reader is implemented based on
    [Databricks statemenet execution API](https://docs.databricks.com/api/workspace/statementexecution).

    Examples:
        import ray
        ds = ray.data.read_databricks_tables(
            warehouse_id='a885ad08b64951ad',
            catalog='catalog_1',
            schema='db_1',
            query='select id from table_1 limit 750000',
        )

    Args:
        warehouse_id: The id of the databricks warehouse, the query statement is
            executed on this warehouse.
        table: The name of UC table you want to read. If this argument is set,
            you can't set 'query' argument, and the reader generates query
            of 'select * from {table_name}' under the hood.
        query: The query you want to execute. If this argument is set,
            you can't set 'table_name' argument.
        catalog: (Optional) The default catalog name used by the query
        schema: (Optional) The default schema used by the query
        parallelism: The requested parallelism of the read. Defaults to -1,
            which automatically determines the optimal parallelism for your
            configuration. You should not need to manually set this value in most cases.
            For details on how the parallelism is automatically determined and guidance
            on how to tune it, see :ref:`Tuning read parallelism
            <read_parallelism>`.
        ray_remote_args: kwargs passed to :meth:`~ray.remote` in the read tasks.

    Returns:
        A :class:`Dataset` containing the queried data.
    """
```
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/datasource/databricks_uc_datasource.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                auth=req_auth,
+                headers=req_headers,
+            )
+            response.raise_for_status()
+            state = response.json()["status"]["state"]
+
+        if state != "SUCCEEDED":
+            raise RuntimeError(f"Query {self.query} execution failed.")
+
+        chunks = response.json()["manifest"]["chunks"]
```
Comment (WeichenXu123) [inline python/ray/data/datasource/databricks_uc_datasource.py:76]: TODO:
if result dataset larger than 100 GiB, the response data is truncated to 100GiB, we need add a warning for the case.

File: python/ray/data/datasource/databricks_uc_datasource.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        state = response.json()["status"]["state"]
+        while state in ["PENDING", "RUNNING"]:
+            time.sleep(1)
+            response = requests.get(
+                urljoin(url_base, statement_id) + "/",
+                auth=req_auth,
+                headers=req_headers,
+            )
+            response.raise_for_status()
+            state = response.json()["status"]["state"]
```
Comment (WeichenXu123) [inline python/ray/data/datasource/databricks_uc_datasource.py:71]: TODO:

During the waiting time, if `KeyboardInterrupt` signal (e.g., in interactive python REPL, user presses Ctrl + C) is received, we should cancel the statement execution by [API](https://docs.databricks.com/api/workspace/statementexecution/cancelexecution) `/api/2.0/sql/statements/{statement_id}/cancel`

File: python/ray/data/datasource/databricks_uc_datasource.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+from ray.data.block import Block, BlockAccessor, BlockMetadata
+from ray.data.datasource.datasource import Datasource, Reader, ReadTask, WriteResult
+from ray.util.annotations import PublicAPI
+
+
+class _DatabricksUCReader(Reader):
+    def __init__(
+        self,
+        host: str,
+        token: str,
```
Comment (WeichenXu123) [inline python/ray/data/datasource/databricks_uc_datasource.py:25]: TODO:

In databricks notebook, we can automatically fill the 2 arguments by:

```
ctx = dbutils.notebook.entry_point.getDbutils().notebook().getContext()
host = ctx.tags().get("browserHostName").get()
token = ctx.apiToken().get()
```

File: python/ray/data/read_api.py [hunk 4]
```diff
@@ -1755,6 +1755,86 @@ def create_connection():
...
 
+@PublicAPI(stability="alpha")
+def read_databricks_uc_tables(
+    *,
+    host: str,
+    token: str,
+    warehouse_id: str,
+    catalog: str,
+    schema: str,
+    table_name: Optional[str] = None,
```
Comment (dbczumar) [inline python/ray/data/read_api.py:1766]: Can we create a single argument called `table`? If users have already set the current catalog / schema in their notebook environment via `USE CATALOG` / `USE SCHEMA`, can we infer the catalog and schema? If it's too tricky, not a big deal for now

File: python/ray/data/read_api.py [hunk 5]
```diff
@@ -1755,6 +1755,86 @@ def create_connection():
     )
 
 
+@PublicAPI(stability="alpha")
+def read_databricks_uc_tables(
```
Comment (dbczumar) [inline python/ray/data/read_api.py:1759]: I like the name for now; in the future, it'll be great to have `read_delta` without `warehouse_id` :D

File: python/ray/data/read_api.py [hunk 6]
```diff
@@ -1755,6 +1755,86 @@ def create_connection():
...
+    *,
+    host: str,
+    token: str,
```
[the remark this answers is not in the document]
Comment (WeichenXu123) [reply python/ray/data/read_api.py:1762]: Yes, see my comment here https://github.com/ray-project/ray/pull/39852#discussion_r1337040266

File: python/ray/data/read_api.py [hunk 7]
```diff
@@ -1755,6 +1755,86 @@ def create_connection():
...
+    warehouse_id: str,
```
[the remark this answers is not in the document]
Comment (WeichenXu123) [reply python/ray/data/read_api.py:1763]: it is already key-work only argument , (defined after `*,`), we can make it optional in future.

File: python/ray/data/read_api.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dbczumar) [reply python/ray/data/read_api.py:1762]: Got it. In that case, can we remove these args for now & users can set the `DATABRICKS_HOST` / `DATABRICKS_TOKEN` env vars if they need to define these?

File: python/ray/data/read_api.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dbczumar) [reply python/ray/data/read_api.py:1763]: 🤦 . Sorry.

File: python/ray/data/read_api.py (hunk 4 above)
Comment (WeichenXu123) [reply python/ray/data/read_api.py:1766]: we can, but I prefer to make `catalog` and `schema` arguments to be optional, in databricks shard, `catalog` and `schema` already has default values, but with the optional arguments, user can easily override them.

File: python/ray/data/datasource/databricks_uc_datasource.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+        req_auth = ('token', self.token)
+        req_headers = {'Content-Type': 'application/json'}
+
+        response = requests.post(
+            url_base,
+            auth=('token', self.token),
+            headers={'Content-Type': 'application/json'},
+            data=payload
+        )
```
Comment (harupy) [inline python/ray/data/datasource/databricks_uc_datasource.py:62]:
```suggestion
        )
        response.raise_for_status()
```

to ensure the request is successful.

File: python/ray/data/datasource/databricks_uc_datasource.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+        try:
+            while state in ["PENDING", "RUNNING"]:
+                time.sleep(1)
+                response = requests.get(
+                    urljoin(url_base, statement_id) + "/",
+                    auth=req_auth,
+                    headers=req_headers,
+                )
+                response.raise_for_status()
+                state = response.json()["status"]["state"]
```
Comment (harupy) [inline python/ray/data/datasource/databricks_uc_datasource.py:77]: no timeout?

File: python/ray/data/datasource/databricks_uc_datasource.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+        except KeyboardInterrupt:
+            # User cancel the command, so we cancel query execution.
+            requests.post(
+                urljoin(url_base, f"{statement_id}/cancel"),
+                auth=req_auth,
+                headers=req_headers,
+            )
+            response.raise_for_status()
```
Comment (harupy) [inline python/ray/data/datasource/databricks_uc_datasource.py:85]: should we print a warning if the cancel fails?

File: python/ray/data/datasource/databricks_uc_datasource.py (hunk 9 above)
Comment (harupy) [reply python/ray/data/datasource/databricks_uc_datasource.py:77]: I see, we don't know how long `query` takes.

File: python/ray/data/datasource/databricks_uc_datasource.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+                        auth=req_auth,
+                        headers=req_headers
+                    )
+                    resolve_response.raise_for_status()
+                    external_url = resolve_response.json()["external_links"][0]["external_link"]
+                    # NOTE: do _NOT_ send the authorization header to external urls
+                    raw_response = requests.get(external_url, auth=None, headers=None)
+                    raw_response.raise_for_status()
+
+                    arrow_table = pyarrow.ipc.open_stream(raw_response.content).read_all()
```
Comment (harupy) [inline python/ray/data/datasource/databricks_uc_datasource.py:148]: Is `raw_response.content` large?

File: python/ray/data/datasource/databricks_uc_datasource.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+                    yield arrow_table
+
+            return ReadTask(read_fn=read_fn, metadata=metadata)
+
+        self._get_read_task = get_read_task
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        return self._estimate_inmemory_data_size
```
Comment (harupy) [inline python/ray/data/datasource/databricks_uc_datasource.py:156]:
```suggestion
    def estimate_inmemory_data_size(self) -> int:
        return self._estimate_inmemory_data_size
```

looks like `self._estimate_inmemory_data_size` always int.

File: python/ray/data/datasource/databricks_uc_datasource.py (hunk 9 above)
Comment (WeichenXu123) [reply python/ray/data/datasource/databricks_uc_datasource.py:77]: We allow user to press Ctrl + C to cancel it, this should fulfil most cases

Comment (WeichenXu123) [reply python/ray/data/datasource/databricks_uc_datasource.py:77]: How long query takes is hard to estimate, it depends on sql workloads, query queues, warehouse resources, etc.

File: python/ray/data/datasource/databricks_uc_datasource.py (hunk 11 above)
Comment (WeichenXu123) [reply python/ray/data/datasource/databricks_uc_datasource.py:148]: usually 10 ~ 20MB size per chunk

File: python/ray/data/datasource/databricks_uc_datasource.py (hunk 12 above)
Comment (WeichenXu123) [reply python/ray/data/datasource/databricks_uc_datasource.py:156]: just to keep the same API signature with base class

Comment (harupy) [reply python/ray/data/datasource/databricks_uc_datasource.py:156]: makes sense

**Comment by WeichenXu123** [comment]
Gentle ping @c21

File: python/ray/data/read_api.py [hunk 13]
```diff
@@ -368,7 +369,12 @@ def read_datasource(
             _get_reader, retry_exceptions=False, num_cpus=0
         ).options(scheduling_strategy=scheduling_strategy)
 
-        (requested_parallelism, min_safe_parallelism, inmemory_size, reader,) = ray.get(
+        (
+            requested_parallelism,
+            min_safe_parallelism,
+            inmemory_size,
+            reader,
+        ) = ray.get(
```
Comment (bveeramani) [inline python/ray/data/read_api.py:377]: Unrelated change?

File: python/ray/data/read_api.py [hunk 14]
```diff
@@ -1755,6 +1761,109 @@ def create_connection():
     )
 
 
+@PublicAPI(stability="alpha")
+def read_databricks_uc_tables(
```
Comment (bveeramani) [inline python/ray/data/read_api.py:1765]: Thoughts on `read_databricks` vs. `read_databricks_tables` vs. `read_databricks_uc_tables`? `read_databricks_uc_tables` sounds slightly verbose to me, but I'm not sure if it'd be misleading if we excluded "tables" or "uc".

File: python/ray/data/read_api.py [hunk 15]
```diff
@@ -1755,6 +1761,109 @@ def create_connection():
...
+    warehouse_id: str,
+    table: Optional[str] = None,
+    query: Optional[str] = None,
+    catalog: Optional[str] = None,
+    schema: Optional[str] = None,
+    parallelism: int = -1,
+    ray_remote_args: Optional[Dict[str, Any]] = None,
+) -> Dataset:
+    """
+    Read from a Databricks UC table or Databricks SQL execution result that queries
```
Comment (bveeramani) [inline python/ray/data/read_api.py:1776]: Explicitly spell out UC since it's the first time this acronym is referenced?

File: python/ray/data/read_api.py [hunk 16]
```diff
@@ -1755,6 +1761,109 @@ def create_connection():
...
+    from Databricks UC tables.
+    If it is not called in databricks runtime, you need to set environment varaibles
+    'DATABRICKS_HOST' and 'DATABRICKS_TOKEN' firstly.
```
Comment (bveeramani) [inline python/ray/data/read_api.py:1779]: Does "it" mean "API" here?
```suggestion
    If this API isn't called in Databricks runtime, set the 'DATABRICKS_HOST' and 'DATABRICKS_TOKEN'
    environment variables.
```

File: python/ray/data/read_api.py [hunk 17]
```diff
@@ -1755,6 +1761,109 @@ def create_connection():
...
+
+    This reader is implemented based on
+    [Databricks statemenet execution API](https://docs.databricks.com/api/workspace/statementexecution).
```
Comment (bveeramani) [inline python/ray/data/read_api.py:1782]:
```suggestion
    [Databricks statement execution API](https://docs.databricks.com/api/workspace/statementexecution).
```

File: python/ray/data/read_api.py [hunk 18]
```diff
@@ -1755,6 +1761,109 @@ def create_connection():
...
+    Examples:
+        >>> from ray.data.read_api import read_databricks_uc_tables
+        >>>
+        >>> ds = read_databricks_uc_tables(
+        ...   warehouse_id='a885ad08b64951ad',
+        ...   catalog='catalog_1',
+        ...   schema='db_1',
+        ...   query='select id from table_1 limit 750000',
+        ...   parallelism=2,
+        ... )
```
Comment (bveeramani) [inline python/ray/data/read_api.py:1793]: * Use code blocks rather than REPL-style examples to improve the readability of this example.
* We don't want to encourage users to manually set `parallelism`. The system-chosen value should (hopefully) work best.
* `ray.data.read_*` is more conventional than importing `read_*` APIs directly.

```suggestion
        .. testcode::
            :skipif: True
            
            import ray
      
            ds = ray.data.read_databricks_uc_tables(
                warehouse_id='a885ad08b64951ad',
                catalog='catalog_1',
                schema='db_1',
                query='select id from table_1 limit 750000',
            )
```

File: python/ray/data/read_api.py [hunk 19]
```diff
@@ -1755,6 +1761,109 @@ def create_connection():
...
+
+    Args:
+        warehouse_id: The id of the databricks warehouse, the query statement is
+            executed on this warehouse.
```
Comment (bveeramani) [inline python/ray/data/read_api.py:1797]:
```suggestion
        warehouse_id: The ID of the Databricks warehouse. The query statement is
            executed on this warehouse.
```

File: python/ray/data/read_api.py [hunk 20]
```diff
@@ -1755,6 +1761,109 @@ def create_connection():
...
+        table: The name of UC table you want to read. If this argument is set,
+            you can't set 'query' argument, and the reader generates query
+            of 'select * from {table_name}' under the hood.
```
Comment (bveeramani) [inline python/ray/data/read_api.py:1800]:
```suggestion
            of ``select * from {table_name}`` under the hood.
```

File: python/ray/data/read_api.py [hunk 21]
```diff
@@ -1755,6 +1761,109 @@ def create_connection():
...
+        ...   schema='db_1',
+        ...   query='select id from table_1 limit 750000',
+        ...   parallelism=2,
+        ... )
+
+    Args:
+        warehouse_id: The id of the databricks warehouse, the query statement is
+            executed on this warehouse.
+        table: The name of UC table you want to read. If this argument is set,
+            you can't set 'query' argument, and the reader generates query
```
Comment (bveeramani) [inline python/ray/data/read_api.py:1799]:
```suggestion
            you can't set ``query`` argument, and the reader generates query
```

File: python/ray/data/read_api.py [hunk 22]
```diff
@@ -1755,6 +1761,109 @@ def create_connection():
...
+            of 'select * from {table_name}' under the hood.
+        query: The query you want to execute. If this argument is set,
+            you can't set 'table_name' argument.
```
Comment (bveeramani) [inline python/ray/data/read_api.py:1802]:
```suggestion
            you can't set ``table_name`` argument.
```

File: python/ray/data/read_api.py [hunk 23]
```diff
@@ -1755,6 +1761,109 @@ def create_connection():
...
+        catalog: (Optional) The default catalog name used by the query
+        schema: (Optional) The default schema used by the query
```
Comment (bveeramani) [inline python/ray/data/read_api.py:1804]:
```suggestion
        catalog: (Optional) The default catalog name used by the query.
        schema: (Optional) The default schema used by the query.
```

File: python/ray/data/read_api.py [hunk 24]
```diff
@@ -1755,6 +1761,109 @@ def create_connection():
...
+            on how to tune it, see :ref:`Tuning read parallelism
+            <read_parallelism>`.
+        ray_remote_args: kwargs passed to :meth:`~ray.remote` in the read tasks.
+
+    Returns:
+        A :class:`Dataset` containing the queried data.
+    """
+    from ray.data.datasource.databricks_uc_datasource import DatabricksUCDatasource
+    from ray.util.spark.databricks_hook import get_dbutils
+    from ray.util.spark.utils import get_spark_session, is_in_databricks_runtime
```
Comment (bveeramani) [inline python/ray/data/read_api.py:1818]: Any reason we can't import this at the top level?

File: python/ray/data/read_api.py [hunk 25]
```diff
@@ -1755,6 +1761,109 @@ def create_connection():
...
+                get_dbutils().notebook.entry_point.getDbutils().notebook().getContext()
+            )
+            if not host:
+                host = ctx.tags().get("browserHostName").get()
+            if not token:
+                token = ctx.apiToken().get()
+        else:
+            raise ValueError(
+                "You are not in databricks runtime, please set environment variable "
+                "'DATABRICKS_HOST' and 'DATABRICKS_TOKEN'."
```
Comment (bveeramani) [inline python/ray/data/read_api.py:1835]: Is there any relevant documentation we could link to? Specifically, how can would a user find `DATABRICKS_HOST` and `DATABRICKS_TOKEN`?

File: python/ray/data/datasource/databricks_uc_datasource.py [hunk 26]
```diff
@@ -0,0 +1,188 @@
...
+        )
+        response.raise_for_status()
+        statement_id = response.json()["statement_id"]
+
+        state = response.json()["status"]["state"]
+
+        logger.info(f"Waiting for query {query} exeuction result.")
+        try:
+            while state in ["PENDING", "RUNNING"]:
+                time.sleep(1)
```
Comment (bveeramani) [inline python/ray/data/datasource/databricks_uc_datasource.py:65]: Nit: Name `1` with a variable? Maybe something like `POLL_TIME_S`? Named constants are good.

File: python/ray/data/datasource/databricks_uc_datasource.py [hunk 27]
```diff
@@ -0,0 +1,188 @@
...
+            try:
+                response.raise_for_status()
+            except Exception as e:
+                logger.warning(
+                    f"Canceling query {query} execution failed, reason: {repr(e)}."
+                )
+            raise
+
+        if state != "SUCCEEDED":
+            raise RuntimeError(f"Query {self.query} execution failed.")
```
Comment (bveeramani) [inline python/ray/data/datasource/databricks_uc_datasource.py:89]: Here and elsewhere so that the query is distinct from the error message
```suggestion
            raise RuntimeError(f"Query {self.query!r} execution failed.")
```

File: python/ray/data/datasource/databricks_uc_datasource.py [hunk 28]
```diff
@@ -0,0 +1,188 @@
...
+            metadata = BlockMetadata(
+                num_rows=num_rows,
+                size_bytes=size_bytes,
+                schema=None,
+                input_files=None,
+                exec_stats=None,
+            )
+
+            def read_fn():
+                for chunk_index in chunk_index_list:
```
Comment (bveeramani) [inline python/ray/data/datasource/databricks_uc_datasource.py:127]: Should we be reading the last chunk only after all of the other chunks have been read? IIRC if you read the last chunk, Databricks marks the statement as closed and you can't access any of the other chunks.

File: python/ray/data/datasource/databricks_uc_datasource.py [hunk 29]
```diff
@@ -0,0 +1,188 @@
...
+
+        manifest = response.json()["manifest"]
+        is_truncated = manifest["truncated"]
+
+        if is_truncated:
+            logger.warning(
+                f"The result dataset of '{query}' exceeding 100GiB and it is truncated."
+            )
+
+        chunks = manifest["chunks"]
```
Comment (bveeramani) [inline python/ray/data/datasource/databricks_uc_datasource.py:99]: The manifest contains the first chunk so we could avoid re-reading that, right? Don't think the additional code complexity is worth the performance improvement, but just wanted to make sure that was considered

**Review by bveeramani — COMMENTED** [review]
Implementation overall LGTM.

Two concerns:
* How do we handle retries, especially for rate limiting errors?
* IIRC Databricks marks a statement as closed when the last statement has been read. Is this still an issue, and if so, do how do we workaround it?

File: python/ray/data/read_api.py (hunk 14 above)
Comment (WeichenXu123) [reply python/ray/data/read_api.py:1765]: I think `read_databricks_tables` is good :)

File: python/ray/data/datasource/databricks_uc_datasource.py (hunk 28 above)
Comment (WeichenXu123) [reply python/ray/data/datasource/databricks_uc_datasource.py:127]: uh ? I don't think we have this limitation. Databricks warehouse caches all chunks data in server side for 1 hour, and we can fetch all chunks in any order and we can fetch chunk data multiple times. Let me confirm this with our databricks warehouse engineers.

File: python/ray/data/datasource/databricks_uc_datasource.py (hunk 29 above)
Comment (WeichenXu123) [reply python/ray/data/datasource/databricks_uc_datasource.py:99]: > The manifest contains the first chunk

No, the manifest only contains metadata of every chunk, but it does not contain first chunk data bytes :)

File: python/ray/data/datasource/databricks_uc_datasource.py (hunk 28 above)
Comment (WeichenXu123) [reply python/ray/data/datasource/databricks_uc_datasource.py:127]: I have confirmed with our databricks warehouse engineer, he says my understanding is correct.
 it used to work like it's described like you said, but not anymore.

File: python/ray/data/read_api.py (hunk 15 above)
Comment (WeichenXu123) [reply python/ray/data/read_api.py:1776]: unity catalog :)

File: python/ray/data/read_api.py (hunk 24 above)
Comment (WeichenXu123) [reply python/ray/data/read_api.py:1818]: it will import pyspark package, I hope we can lazily import `pyspark` , importing pyspark package is slow.

File: python/ray/data/read_api.py (hunk 25 above)
Comment (WeichenXu123) [reply python/ray/data/read_api.py:1835]: It is already in this API doc.

**Comment by WeichenXu123** [comment]
@bveeramani 

I got replied from our warehouse engineer about rate limit:

For concurrent request (concurrent fetching chunk data) limiters of EXTERNAL_LINKS type (the type used in my PR), it's around 750 requests per second per workspace.


So I think we rarely hit the rate limit , because one chunk data is at least 10MB size and each request takes some time to complete, if hit rate limit, we can just let ray task retry.

File: python/ray/data/read_api.py [hunk 30]
```diff
@@ -1755,6 +1756,114 @@ def create_connection():
...
+    from ray.data.datasource.databricks_uc_datasource import DatabricksUCDatasource
+    from ray.util.spark.databricks_hook import get_dbutils
+    from ray.util.spark.utils import get_spark_session, is_in_databricks_runtime
+
+    token = os.environ.get("DATABRICKS_TOKEN")
+
+    if not token:
+        raise ValueError(
+            "Please set environment variable 'DATABRICKS_TOKEN'."
+        )
```
Comment (WeichenXu123) [inline python/ray/data/read_api.py:1823]: Update:

In databricks notebook, we still need user to set 'DATABRICKS_TOKEN' env variable, the builtin token does not work for databricks warehouse.

File: python/ray/data/datasource/databricks_uc_datasource.py (hunk 28 above)
Comment (bveeramani) [reply python/ray/data/datasource/databricks_uc_datasource.py:131]: Gotcha. That's awesome

File: python/ray/data/datasource/databricks_uc_datasource.py (hunk 29 above)
Comment (bveeramani) [reply python/ray/data/datasource/databricks_uc_datasource.py:103]: Ah, I meant does `manfiest["result"]` contain the external links for the first chunk? In any case, it's probably not worth the complexity of this sort of optimization even if it's possible

File: python/ray/data/read_api.py (hunk 24 above)
Comment (bveeramani) [reply python/ray/data/read_api.py:1816]: Gotcha, makes sense

File: python/ray/data/read_api.py (hunk 25 above)
Comment (bveeramani) [reply python/ray/data/read_api.py]: As in the docstring? I think we just link to https://docs.databricks.com/api/workspace/statementexecution, which I don't think covers how to find the environment variables?

Also, to make the error message more actionable, I think it'd be great if we could say something like:

"Please set environment variable ... To get these values, see [relevant Databricks page]"

File: python/ray/data/read_api.py [hunk 31]
```diff
@@ -1755,6 +1756,114 @@ def create_connection():
...
+        :skipif: True
+
+        import ray
+
+        ds = ray.data.read_databricks_tables(
+            warehouse_id='a885ad08b64951ad',
+            catalog='catalog_1',
+            schema='db_1',
+            query='select id from table_1 limit 750000',
+        )
```
Comment (bveeramani) [inline python/ray/data/read_api.py:1791]: In addition to this example here, could you update the "Databricks" section of "loading-data.rst"?

**Review by bveeramani — APPROVED** [review]
Awesome! Overall LGTM!

Once we've resolved the remaining review comments, this should be good to merge.

> For concurrent request (concurrent fetching chunk data) limiters of EXTERNAL_LINKS type (the type used in my PR), it's around 750 requests per second per workspace.

Got it. Think this should be okay in most cases, but if you have a cluster with more than 750 cores I can see this being an issue. IIRC we ran into some rate limiting errors when we performed internal testing.

In any case, I think we can defer this for now and address this as a follow up if people run into issues.

> if hit rate limit, we can just let ray task retry.

Ray Tasks don't retry application-level exceptions by default. So, if you get a rate limiting error,  I don't think it'd get retried.

File: python/ray/data/read_api.py (hunk 31 above)
Comment (WeichenXu123) [reply python/ray/data/read_api.py:1791]: Added! :)

File: python/ray/data/read_api.py (hunk 25 above)
Comment (WeichenXu123) [reply python/ray/data/read_api.py]: Oh I got it, these 2 environment variable keys are defined in `read_databricks_tables` code, so we don't have databricks doc for them.

I refined `read_databricks_tables` API doc to add description for the 2 environment variables:
'DATABRICKS_HOST' means databricks workspace URL, like "adb-<workspace-id>.<random-number>.azuredatabricks.net",
'DATABRICKS_TOKEN' means databricks workspace access token, user knows how to get it if it saw the doc description.

**Comment by WeichenXu123** [comment]
> if you have a cluster with more than 750 cores I can see this being an issue. IIRC we ran into some rate limiting errors when we performed internal testing.

> Ray Tasks don't retry application-level exceptions by default. So, if you get a rate limiting error, I don't think it'd get retried.


We can address them in follow-up PRs, since this case rarely happens

**Comment by bveeramani** [comment]
> We can address them in follow-up PRs, since this case rarely happens

Sounds good. Thanks for your work on this PR!

**Comment by bveeramani** [comment]
@WeichenXu123 When you're back from vacation, could add tests? My bad -- I should've caught this in an earlier review and before merging

**Comment by WeichenXu123** [comment]
> @WeichenXu123 When you're back from vacation, could add tests? My bad -- I should've caught this in an earlier review and before merging

The e2e test requires databricks environment, we will add e2e test in our databricks code repo to monitor the Ray reader works.

For unit test in Ray repo, we can only add mocking test, we can add it if you need it.

**Comment by bveeramani** [comment]
@WeichenXu123 ah, gotcha! Didn't realize you were running E2E tests on your side.

Given that we'd likely need extensive mocks to unit tests this feature, I think the additional complexity may outweigh the benefit of the tests.
